### PR TITLE
FIX #78 : try to avoid (UnhandledPromiseRejectionWarning)

### LINKED ISSUE #78: error handling
What's the recommended way of handling errors inside (`async`) plugins?

Given an example plugin like the following:

``` javascript
async function plugin() {
  // uncomment next line to throw error in plugin initialization phase
  // throw new Error("plugin INIT error")

  return async files => {
    throw new Error("plugin EXEC error")
  }
}
```

I'll get an `UnhandledPromiseRejectionWarning` in both error cases. These are notoriously hard to debug. Should these errors be caught by awoo or the plugin?

Still getting confused with error handling when using promises&hellip;

Thanks a bunch  :octocat:

### PR DESCRIPTION
**Pull request by zipang**
Hi @oe !
Really like this project, as i had for some times now the same idea : basically rewriting Metalsmith with a cleaner API 👍 
Coming from Metalsmith can be hard because of this nasty exception.

Anyway.., this PR really is not satisfactory...
But it should give newcomers a better look at what they are doing wrong.

### CODE REVIEW

**Comment by oe** [comment]
Thanks for this! I'll sort out the bad and weird CI failures.

File: lib/awoo.js [hunk 1]
```diff
@@ -5,8 +5,14 @@ const write = require('./write')
 const { generateConfig } = require('./config')
 
 async function siteUse (fn, opts = {}) {
-  const plugin = await fn(opts)
-  this.hooks.use(plugin)
+  try {
+    const plugin = await fn(opts)
+    this.hooks.use(plugin)
+  } catch(err) {
+    debug(`plugin ${fn.name ||fn} init failed :`)
```
Comment (oe) [inline lib/awoo.js]: Could you remove the space between the end of `failed` and the colon?

File: lib/awoo.js [hunk 2]
```diff
@@ -5,8 +5,14 @@ const write = require('./write')
...
+    debug(err);
+    process.exit(1); // don't continue
```
Comment (oe) [inline lib/awoo.js]: Would be cool to also just throw the error publicly (i.e. not only in debug) if we just hard exit here.

File: lib/awoo.js [hunk 3]
```diff
@@ -35,8 +41,15 @@ async function buildSite (site) {
...
 
-  debug('running plugins...')
-  site.files = await _runMiddleware(site.hooks, site.files)
+  try {
+    debug('running plugins...')
+    site.files = await _runMiddleware(site.hooks, site.files)
+  } catch (err) {
+    debug('an error occured while applying plugins:')
+    debug(err)
+    debug("ABORTING..")
```
Comment (oe) [inline lib/awoo.js]: Don't think we need the ABORTING line here.

File: lib/awoo.js (hunk 1 above)
Comment (oe) [reply lib/awoo.js]: Oh yeah, and add a space so it looks like `fn.name || fn`?

File: lib/awoo.js [hunk 4]
```diff
@@ -35,8 +41,15 @@ async function buildSite (site) {
...
+    process.exit(1); // don't continue
```
Comment (oe) [inline lib/awoo.js]: Same as above with regards to throwing errors!

File: lib/awoo.js [hunk 5]
```diff
@@ -56,22 +69,22 @@ async function awooCore (site, fn) {
...
+  if (Object.keys(finalSite._config).length === 0) {
+    finalSite._config = generateConfig({})
   }
-  return buildSite(resSite)
+  return buildSite(finalSite)
 }
 
 async function awoo (fn) {
-  const site = {}
+  var site = {}
```
Comment (oe) [inline lib/awoo.js]: Was something about this wrong? Assigning properties doesn't need a `var` statement, right?

File: lib/awoo.js [hunk 6]
```diff
@@ -56,22 +69,22 @@ async function awooCore (site, fn) {
 
   debug('awoo is ready to go!!!')
 
-  const resSite = await fn(site)
-  if (Object.keys(site._config).length === 0) {
-    site._config = generateConfig({})
+  const finalSite = await fn(site) || site
+  if (Object.keys(finalSite._config).length === 0) {
```
Comment (oe) [inline lib/awoo.js]: good catch!

**Comment by oe** [comment]
This will hopefully close #78!

File: lib/awoo.js (hunk 5 above)
Comment (oe) [reply lib/awoo.js]: Oh, I think this'll be fine.

File: lib/awoo.js (hunk 2 above)
Comment (zipang) [reply lib/awoo.js]: yep... [debug()](https://github.com/visionmedia/debug) really doesn't have a notion of _different levels_ of debugging.. unlike [winston](https://github.com/winstonjs/winston) for instance : It's really only a switch on/off
So, what should we use instead ? console.err() ?

Comment (oe) [reply lib/awoo.js]: Yeah, that'll work.

**Comment by oe** [comment]
@zipang It's fine, I'll merge this and then take care of it. Thanks so much!

**Pull request merged**